File: quickscope.cpp
#include <bits/stdc++.h>
#include <string>
#include <unordered_map>

using namespace std;

typedef vector<int> vi;
typedef unordered_map<string,string> ss;
typedef unordered_map<string,vi> svi;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    int N; cin >> N;
    vector<ss> stack; svi types; ss curr;
    for (int i = 0; i < N; i++) {
        string a; cin >> a;
        if (a == "{") {
            stack.push_back(curr);
            curr = ss ();
        } else if (a == "}") {
            ss prev = curr;
            curr = stack.back(); stack.pop_back();
            for (auto &[key,value] : prev) {
                if (types[key].back() == stack.size() + 1) {
                    types[key].pop_back();
                }
            }
        } else if (a == "DECLARE") {
            string v,t; cin >> v >> t;
            if (curr.find(v) != curr.end()) {
                cout << "MULTIPLE DECLARATION\n";
                break;
            }
            curr[v] = t;
            if (types.find(v) != types.end()) {
                types[v].push_back((int) stack.size());
            } else {
                types[v] = vi {(int) stack.size()};
            }
        } else { //TYPEOF query
            string v; cin >> v;
            if (types.find(v) == types.end() || types[v].empty()) cout << "UNDECLARED\n";
            else {
                int lastIndex = types[v].back();
                cout << (lastIndex >= stack.size() ? curr[v] : stack[lastIndex][v])<< "\n";
            }
        }
    }
    return 0;
}

File: quickscope.py
from sys import stdin,stdout

#data structures used:
#stack of hashmaps mapping variables to their types 
#hashmaps mapping variables to stacks of their indices
stack,types,curr = [],{},{}
N = int(stdin.readline())
for _ in range(N):
    line = stdin.readline().strip().split(" ")
    if line[0] == "{":
        stack.append(curr)
        curr = {}
    elif line[0] == "}":
        prev = curr
        curr = stack.pop()
        for key in prev:
            if types[key][-1] == len(stack) + 1:
                types[key].pop()
    elif line[0] == "DECLARE":
        v,t = line[1],line[2]
        if v in curr:
            stdout.write("MULTIPLE DECLARATION\n")
            break
        curr[v] = t
        if v not in types:
            types[v] = [len(stack)]
        else:
            types[v].append(len(stack))
    else: #TYPEOF query
        v = line[1]
        if v not in types or not types[v]:
            stdout.write("UNDECLARED\n")
        else:
            lastIndex = types[v][-1]
            stdout.write(f"{curr[v]}\n") if lastIndex >= len(stack) else stdout.write(f"{stack[lastIndex][v]}\n")
            
            
        
